actionview/lib/action_view/helpers/translation_helper.rb

```diff
@@ -58,36 +58,38 @@ module TranslationHelper
       # when you call translate in a template and translators know which keys
       # they can provide HTML values for.
       def translate(key, **options)
-        unless options[:default].nil?
-          remaining_defaults = Array.wrap(options.delete(:default)).compact
-          options[:default] = remaining_defaults unless remaining_defaults.first.kind_of?(Symbol)
-        end
+        return key.map { |k| translate(k, **options) } if key.is_a?(Array)
 
-        # If the user has explicitly decided to NOT raise errors, pass that option to I18n.
-        # Otherwise, tell I18n to raise an exception, which we rescue further in this method.
-        # Note: `raise_error` refers to us re-raising the error in this method. I18n is forced to raise by default.
-        if options[:raise] == false
-          raise_error = false
-          i18n_raise = false
-        else
-          raise_error = options[:raise] || ActionView::Base.raise_on_missing_translations
-          i18n_raise = true
+        defaults = if options.key?(:default)
+          options[:default].is_a?(Array) ? options[:default].compact : [options[:default]]
         end
 
-        if html_safe_translation_key?(key)
-          html_safe_options = html_escape_translation_options(options)
-          translation = I18n.translate(scope_key_by_partial(key), **html_safe_options.merge(raise: i18n_raise))
-          html_safe_translation(translation)
-        else
-          I18n.translate(scope_key_by_partial(key), **options.merge(raise: i18n_raise))
-        end
-      rescue I18n::MissingTranslationData => e
-        if remaining_defaults.present?
-          translate remaining_defaults.shift, **options.merge(default: remaining_defaults)
-        else
-          raise e if raise_error
-          missing_translation(e, options)
+        options[:raise] = true if options[:raise].nil? && ActionView::Base.raise_on_missing_translations
+        sentinel = MISSING_TRANSLATION
+
+        while key
+          if (defaults.nil? || defaults.empty?) && !options[:raise].nil?
+            sentinel = [].freeze # let I18n handle missing translation
```

Comment:
This is a bit confusing: the empty array isn't a sentinel value, it's telling I18n not to return a default so that the `raise` option takes effect. The clearest way to do this would be to omit the `default` option entirely.

I was going to suggest moving this value inside `options`:

```patch
diff --git a/actionview/lib/action_view/helpers/translation_helper.rb b/actionview/lib/action_view/helpers/translation_helper.rb
index 76beeca7a3..67a8f3e66c 100644
--- a/actionview/lib/action_view/helpers/translation_helper.rb
+++ b/actionview/lib/action_view/helpers/translation_helper.rb
@@ -65,11 +65,11 @@ def translate(key, **options)
         end

         options[:raise] = true if options[:raise].nil? && ActionView::Base.raise_on_missing_translations
-        sentinel = MISSING_TRANSLATION
+        options[:default] = MISSING_TRANSLATION

         while key
           if (defaults.nil? || defaults.empty?) && !options[:raise].nil?
-            sentinel = [].freeze # let I18n handle missing translation
+            options.delete(:default) # let I18n handle missing translation
           end

           key = scope_key_by_partial(key)
@@ -77,11 +77,11 @@ def translate(key, **options)

           if html_safe_translation_key?(key)
             html_safe_options ||= html_escape_translation_options(options)
-            translation = I18n.translate(key, **html_safe_options, default: sentinel)
-            return html_safe_translation(translation) unless translation.equal?(sentinel)
+            translation = I18n.translate(key, **html_safe_options)
+            return html_safe_translation(translation) unless translation.equal?(MISSING_TRANSLATION)
           else
-            translation = I18n.translate(key, **options, default: sentinel)
-            return translation unless translation.equal?(sentinel)
+            translation = I18n.translate(key, **options)
+            return translation unless translation.equal?(MISSING_TRANSLATION)
           end

           return defaults.first if defaults && !defaults.empty? && !defaults.first.is_a?(Symbol)
```

Until I realised it introduced a bug that wasn't covered by tests:

```ruby
  def test_translate_with_html_key_and_missing_default_and_raise_option
    assert_raise(I18n::MissingTranslationData) do
      translate(:'translations.missing_html', default: :'translations.missing_html', raise: true)
    end
  end
```

I'm still not sure how to make this any clearer, but would you mind adding that test coverage?

Reply:
I added the test, and changed `translation.equal?(sentinel)` to `translation.equal?(MISSING_TRANSLATION)` for clarity.  Also, I know it's not an enormous difference, but I changed `sentinel` to `guard`.  To me, `guard` is a bit less concrete, and thus more flexible in its interpretation and usage.  What do you think?

Reply:
I think `guard` is a bit better, even if only because it's less intimidating!